File: RepoTransBench/target_projects/Python/Go/mattupstate_overholt/tests/original/middleware_test.go
package original

import (
	"testing"
	"bytes"
	"strings"
	"reflect"
	"mattupstate_overholt/overholt/middleware"
)

// DummyApp in Go, simulating WSGI app interface.
type DummyApp struct {
	LastEnviron      map[string]interface{}
	LastStartStatus  string
	LastStartHeaders [][2]string
}

func (d *DummyApp) ServeHTTP(environ map[string]interface{}, startResponse func(string, [][2]string)) [][]byte {
	d.LastEnviron = environ
	startResponse("200 OK", [][2]string{{"Content-Type", "text/plain"}})
	return [][]byte{[]byte("response")}
}

func makeEnviron(method string, queryString string, headers map[string]string) map[string]interface{} {
	environ := map[string]interface{}{
		"REQUEST_METHOD": method,
		"QUERY_STRING":   queryString,
	}
	for k, v := range headers {
		environ[k] = v
	}
	return environ
}

// Emulates middleware function in Go, assumes similar interface. You must have an implementation matching API.
func TestMethodOverrideHeader(t *testing.T) {
	app := &DummyApp{}
	mw := middleware.NewHTTPMethodOverrideMiddleware(app, "", "", []string{})
	environ := makeEnviron("POST", "", map[string]string{"HTTP_X_HTTP_METHOD_OVERRIDE": "DELETE"})
	called := struct {
		status  string
		headers [][2]string
	}{}
	startResponse := func(status string, headers [][2]string) {
		called.status = status
		called.headers = headers
	}
	result := mw.ServeHTTP(environ, startResponse)
	// In Go: let's assume the middleware modifies REQUEST_METHOD as needed
	if rm, ok := app.LastEnviron["REQUEST_METHOD"]; ok {
		// We expect middleware to override to "DELETE" (as []byte or string)
		switch v := rm.(type) {
		case string:
			if v != "DELETE" {
				t.Errorf("Expected overridden REQUEST_METHOD 'DELETE', got '%v'", v)
			}
		default:
			t.Errorf("REQUEST_METHOD has unexpected type %T", v)
		}
	} else {
		t.Errorf("REQUEST_METHOD missing from environ")
	}
	// We don't have Content-Length in our dummy implementation, but assert result and status
	if !reflect.DeepEqual(result, [][]byte{[]byte("response")}) {
		t.Errorf("Incorrect result: got %v", result)
	}
	if called.status != "200 OK" {
		t.Errorf("Expected status 200 OK, got %s", called.status)
	}
}

func TestMethodOverrideQuerystring(t *testing.T) {
	app := &DummyApp{}
	mw := middleware.NewHTTPMethodOverrideMiddleware(app, "", "", []string{})
	environ := makeEnviron("POST", "foo=bar&__METHOD__=PUT", map[string]string{})
	called := struct {
		status  string
		headers [][2]string
	}{}
	startResponse := func(status string, headers [][2]string) {
		called.status = status
		called.headers = headers
	}
	result := mw.ServeHTTP(environ, startResponse)
	if rm, ok := app.LastEnviron["REQUEST_METHOD"]; ok {
		switch v := rm.(type) {
		case string:
			if v != "PUT" {
				t.Errorf("Expected overridden REQUEST_METHOD 'PUT', got '%v'", v)
			}
		default:
			t.Errorf("REQUEST_METHOD has unexpected type %T", v)
		}
	} else {
		t.Errorf("REQUEST_METHOD missing from environ")
	}
	if !reflect.DeepEqual(result, [][]byte{[]byte("response")}) {
		t.Errorf("Incorrect result: got %v", result)
	}
}

func TestNoOverride(t *testing.T) {
	app := &DummyApp{}
	mw := middleware.NewHTTPMethodOverrideMiddleware(app, "", "", []string{})
	environ := makeEnviron("GET", "", map[string]string{})
	called := struct {
		status  string
		headers [][2]string
	}{}
	startResponse := func(status string, headers [][2]string) {
		called.status = status
		called.headers = headers
	}
	result := mw.ServeHTTP(environ, startResponse)
	if rm, ok := app.LastEnviron["REQUEST_METHOD"]; ok {
		if rm != "GET" {
			t.Errorf("Expected REQUEST_METHOD 'GET', got '%v'", rm)
		}
	} else {
		t.Errorf("REQUEST_METHOD missing from environ")
	}
	if !reflect.DeepEqual(result, [][]byte{[]byte("response")}) {
		t.Errorf("Incorrect result: got %v", result)
	}
}

func TestOverrideWithCustom(t *testing.T) {
	app := &DummyApp{}
	mw := middleware.NewHTTPMethodOverrideMiddleware(app, "X-MY-HEADER", "__MY_METHOD__", []string{"PUT"})
	environ := makeEnviron("POST", "", map[string]string{"HTTP_X_MY_HEADER": "PUT"})
	called := struct {
		status  string
		headers [][2]string
	}{}
	startResponse := func(status string, headers [][2]string) {
		called.status = status
		called.headers = headers
	}
	result := mw.ServeHTTP(environ, startResponse)
	if rm, ok := app.LastEnviron["REQUEST_METHOD"]; ok {
		if rm != "PUT" {
			t.Errorf("Expected overridden REQUEST_METHOD 'PUT', got '%v'", rm)
		}
	} else {
		t.Errorf("REQUEST_METHOD missing from environ")
	}
	if !reflect.DeepEqual(result, [][]byte{[]byte("response")}) {
		t.Errorf("Incorrect result: got %v", result)
	}
}

func TestOverrideNotAllowed(t *testing.T) {
	app := &DummyApp{}
	mw := middleware.NewHTTPMethodOverrideMiddleware(app, "", "", []string{"POST"})
	environ := makeEnviron("POST", "", map[string]string{"HTTP_X_HTTP_METHOD_OVERRIDE": "PATCH"})
	called := struct {
		status  string
		headers [][2]string
	}{}
	startResponse := func(status string, headers [][2]string) {
		called.status = status
		called.headers = headers
	}
	result := mw.ServeHTTP(environ, startResponse)
	if rm, ok := app.LastEnviron["REQUEST_METHOD"]; ok {
		if rm != "POST" {
			t.Errorf("Expected REQUEST_METHOD to stay 'POST', got '%v'", rm)
		}
	} else {
		t.Errorf("REQUEST_METHOD missing from environ")
	}
	if !reflect.DeepEqual(result, [][]byte{[]byte("response")}) {
		t.Errorf("Incorrect result: got %v", result)
	}
}

func TestGetFromQuerystringReturnsNone(t *testing.T) {
	app := &DummyApp{}
	mw := middleware.NewHTTPMethodOverrideMiddleware(app, "", "", []string{})
	environ := makeEnviron("POST", "foo=bar", map[string]string{})
	method := mw.GetFromQuerystring(environ)
	if method != "" {
		t.Errorf("Expected no override from querystring, got %v", method)
	}
}

File: RepoTransBench/target_projects/Python/Java/mattupstate_overholt/src/test/java/com/overholt/original/MiddlewareTest.java
package com.overholt.original;

import com.overholt.middleware.HTTPMethodOverrideMiddleware;
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;
import java.util.*;

class DummyAppImpl implements HTTPMethodOverrideMiddleware.DummyApp {
    private Map<String, Object> lastEnviron;
    private List<Map.Entry<String, String>> lastHeaders;

    @Override
    public List<byte[]> call(Map<String, Object> environ, HTTPMethodOverrideMiddleware.StartResponse startResponse) {
        this.lastEnviron = environ;
        this.lastHeaders = new ArrayList<>();
        startResponse.call("200 OK", List.of(Map.entry("Content-Type", "text/plain")));
        return List.of("response".getBytes());
    }

    @Override
    public Map<String, Object> getLastEnviron() {
        return lastEnviron;
    }

    @Override
    public List<Map.Entry<String, String>> getLastHeaders() {
        return lastHeaders;
    }
}

public class MiddlewareTest {

    private Map<String, Object> makeEnviron(String method, String queryString, Map<String, String> headers) {
        Map<String, Object> environ = new HashMap<>();
        environ.put("REQUEST_METHOD", method);
        environ.put("QUERY_STRING", queryString == null ? "" : queryString);
        if (headers != null)
            environ.putAll(headers);
        return environ;
    }

    @Test
    void testMethodOverrideHeader() {
        DummyAppImpl app = new DummyAppImpl();
        HTTPMethodOverrideMiddleware middleware = new HTTPMethodOverrideMiddleware(app);
        Map<String, Object> environ = makeEnviron("POST", "", Map.of("HTTP_X_HTTP_METHOD_OVERRIDE", "DELETE"));
        final List<String> called = new ArrayList<>();
        HTTPMethodOverrideMiddleware.StartResponse startResp = (status, headers) -> called.add(status);
        List<byte[]> result = middleware.call(environ, startResp);

        Object reqMethod = app.getLastEnviron().get("REQUEST_METHOD");
        assertTrue(reqMethod instanceof byte[] && Arrays.equals((byte[])reqMethod, "DELETE".getBytes()));
        assertTrue(app.getLastEnviron().containsKey("CONTENT_LENGTH"));
        assertEquals(1, result.size());
        assertEquals("response", new String(result.get(0)));
        assertEquals("200 OK", called.get(0));
    }

    @Test
    void testMethodOverrideQuerystring() {
        DummyAppImpl app = new DummyAppImpl();
        HTTPMethodOverrideMiddleware middleware = new HTTPMethodOverrideMiddleware(app);
        Map<String, Object> environ = makeEnviron("POST", "foo=bar&__METHOD__=PUT", null);
        final List<String> called = new ArrayList<>();
        HTTPMethodOverrideMiddleware.StartResponse startResp = (status, headers) -> called.add(status);
        List<byte[]> result = middleware.call(environ, startResp);

        Object reqMethod = app.getLastEnviron().get("REQUEST_METHOD");
        assertTrue(reqMethod instanceof byte[] && Arrays.equals((byte[])reqMethod, "PUT".getBytes()));
        assertTrue(app.getLastEnviron().containsKey("CONTENT_LENGTH"));
        assertEquals(1, result.size());
        assertEquals("response", new String(result.get(0)));
    }

    @Test
    void testNoOverride() {
        DummyAppImpl app = new DummyAppImpl();
        HTTPMethodOverrideMiddleware middleware = new HTTPMethodOverrideMiddleware(app);
        Map<String, Object> environ = makeEnviron("GET", "", null);
        final List<String> called = new ArrayList<>();
        HTTPMethodOverrideMiddleware.StartResponse startResp = (status, headers) -> called.add(status);
        List<byte[]> result = middleware.call(environ, startResp);

        Object reqMethod = app.getLastEnviron().get("REQUEST_METHOD");
        assertTrue(reqMethod instanceof String && reqMethod.equals("GET"));
        assertEquals(1, result.size());
        assertEquals("response", new String(result.get(0)));
    }

    @Test
    void testOverrideWithCustom() {
        DummyAppImpl app = new DummyAppImpl();
        HTTPMethodOverrideMiddleware middleware = new HTTPMethodOverrideMiddleware(
            app, "X-MY-HEADER", "__MY_METHOD__", List.of("PUT")
        );
        Map<String, Object> environ = makeEnviron("POST", "", Map.of("HTTP_X_MY_HEADER", "PUT"));
        final List<String> called = new ArrayList<>();
        HTTPMethodOverrideMiddleware.StartResponse startResp = (status, headers) -> called.add(status);
        List<byte[]> result = middleware.call(environ, startResp);

        Object reqMethod = app.getLastEnviron().get("REQUEST_METHOD");
        assertTrue(reqMethod instanceof byte[] && Arrays.equals((byte[])reqMethod, "PUT".getBytes()));
    }

    @Test
    void testOverrideNotAllowed() {
        DummyAppImpl app = new DummyAppImpl();
        HTTPMethodOverrideMiddleware middleware = new HTTPMethodOverrideMiddleware(app, "X-HTTP-METHOD-OVERRIDE", "__METHOD__", Set.of("POST"));
        Map<String, Object> environ = makeEnviron("POST", "", Map.of("HTTP_X_HTTP_METHOD_OVERRIDE", "PATCH"));
        final List<String> called = new ArrayList<>();
        HTTPMethodOverrideMiddleware.StartResponse startResp = (status, headers) -> called.add(status);
        List<byte[]> result = middleware.call(environ, startResp);

        Object reqMethod = app.getLastEnviron().get("REQUEST_METHOD");
        assertTrue(reqMethod instanceof String && reqMethod.equals("POST"));
    }

    @Test
    void testGetFromQuerystringReturnsNone() {
        DummyAppImpl app = new DummyAppImpl();
        HTTPMethodOverrideMiddleware middleware = new HTTPMethodOverrideMiddleware(app);
        Map<String, Object> environ = makeEnviron("POST", "foo=bar", null);
        assertNull(middleware._getFromQuerystring(environ));
    }
}